src/scattdata.cpp

```diff
@@ -0,0 +1,936 @@
+      int gout = i_gout + gmin[gin];
+      for (int l = 0; l < order_dim; l++) {
+        matrix[gin][gout][l] = scattxs[gin] * energy[gin][i_gout] *
+             fmu[gin][i_gout][l];
+      }
+    }
+  }
+  return matrix;
+}
+
+//==============================================================================
+
+void
+ScattDataHistogram::combine(const std::vector<ScattData*>& those_scatts,
+                            const double_1dvec& scalars)
+{
+  // Find the max order in the data set and make sure we can combine the sets
+  int max_order = those_scatts[0]->get_order();
+  for (int i = 0; i < those_scatts.size(); i++) {
+    // Lets also make sure these items are combineable
+    ScattDataHistogram* that = dynamic_cast<ScattDataHistogram*>(those_scatts[i]);
+    if (!that) {
+      fatal_error("Cannot combine the ScattData objects!");
+    }
+    if (max_order != that->get_order()) {
+      fatal_error("Cannot combine the ScattData objects!");
+    }
+  }
+
+  int groups = those_scatts[0] -> energy.size();
+
+  int_1dvec in_gmin(groups);
+  int_1dvec in_gmax(groups);
+  double_3dvec sparse_scatter(groups);
+  double_2dvec sparse_mult(groups);
+
+  // The rest of the steps do not depend on the type of angular representation
+  // so we use a base class method to sum up xs and create new energy and mult
+  // matrices
+  ScattData::base_combine(max_order, those_scatts, scalars, in_gmin, in_gmax,
+                             sparse_mult, sparse_scatter);
+
+  // Got everything we need, store it.
+  init(in_gmin, in_gmax, sparse_mult, sparse_scatter);
+}
+
+//==============================================================================
+// ScattDataTabular methods
+//==============================================================================
+
+void
+ScattDataTabular::init(const int_1dvec& in_gmin, const int_1dvec& in_gmax,
+     const double_2dvec& in_mult, const double_3dvec& coeffs)
+{
+  int groups = coeffs.size();
+  int order = coeffs[0][0].size();
+
+  // make a copy of coeffs that we can use to both extract data and normalize
+  double_3dvec matrix = coeffs;
+
+  // Build the angular distribution mu values
+  mu = double_1dvec(order);
+  dmu = 2. / (order - 1);
+  mu[0] = -1.;
+  for (int imu = 1; imu < order - 1; imu++) {
+    mu[imu] = -1. + imu * dmu;
+  }
+  mu[order - 1] = 1.;
+
+  // Get the scattering cross section value by integrating the distribution
+  // over all mu points and then combining over all outgoing groups
+  scattxs.resize(groups);
+  for (int gin = 0; gin < groups; gin++) {
+    scattxs[gin] = 0.;
+    for (int i_gout = 0; i_gout < matrix[gin].size(); i_gout++) {
+      for (int imu = 1; imu < order; imu++) {
+        scattxs[gin] += 0.5 * dmu * (matrix[gin][i_gout][imu - 1] +
+                                     matrix[gin][i_gout][imu]);
+      }
+    }
+  }
+
+  // Build the energy transfer matrix from data in the variable matrix
+  double_2dvec in_energy;
+  in_energy.resize(groups);
```

Comment:
`double_2dvec in_energy(groups);`

Reply:
- [x] (I'm seeing my evolution of C++ experience in this module, I started here and went up the code tree)